vshard/replicaset.lua

```diff
@@ -144,6 +144,45 @@ local function netbox_wait_connected(conn, timeout)
     return timeout
 end
 
+--
+-- Check if the connection is dead and it won't be restored automatically.
+-- Even though replicaset's connections are initialized with the option
+-- 'reconnect_after', there's cases, when no reconnect will be done (e.g.
+-- if the conn was explicitly cancelled or the connection's worker fiber
+-- was killed). In these situations we need to reestablish it manually as
+-- the missing connection to some replicaset is a critical problem which
+-- leads to the inability of the user to access some part of a data via
+-- router.
+--
+-- Reconnecting after explicit fiber kill doesn't happen in tarantool 2.10.0
+-- as there's no way to determine if the fiber is dead other than checking
+-- the error message of the connection, which is not a good practice as this
+-- check can be easily trigerred false-positively.
```

Comment:
It works on 2.10.0 too - you check for `conn._fiber:status() == 'dead'`. I run your test on 2.10.0 and it passed. I also tested it in console just in case:
```Lua
netbox = require('net.box')
box.cfg{listen = 3313}
box.schema.user.grant('guest', 'super')
c = netbox.connect(box.cfg.listen, {reconnect_after = 0.1})
```
```
tarantool> c._fiber:cancel()
---
...

tarantool> 2022-08-12 23:54:11.491 [18782] main/118/:3313 (net.box) net_box.lua:307 W> :3313: fiber is cancelled

---
...

tarantool> c._fiber:status()
---
- dead
...
```
I suggest to enable the new test on all versions, remove mentioning of 2.10.0 from the commit message, and from the comment.

Reply:
Surprisingly, you are right. But the commit https://github.com/tarantool/tarantool/commit/0d944f901cf5400d5832a5528780bb3ecd03ae4b, which introduced `_fiber` in `conn`, was merged on June 22nd, which is later than any release of the tarantool 2.10.0.

I used for testing the version, which is tagged with 2.10.0. In CI, however, AFAICS the versions from tarantool's branches are used. I suppose I didn't get it right how the different versions are formed and why do we still push to them after the release. I'll try to figure that out. 

Fixed everything according to your comment.